### PR TITLE
Additional changes to delegate bitcoin core keys to eclair

### PR DESCRIPTION
**Pull request by t-bast**
This is a PR on top of #2613 that fixes all the comments I had while reviewing and testing that code.

Sorry if some comments seem minor (capitalization, scala style nits, etc), that really helped my process of tinkering with the code and spending enough time staring at it to make sure there was no issue!

Each commit is independent and contains its own description, make sure to review commit-by-commit. Some of those commits seem to introduce diff, but they're actually reverting changes from #2613 which will make the overall diff with `master` smaller.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala [hunk 1]
```diff
@@ -125,4 +125,30 @@ object OnChainWallet {
...
   }
+
+  final case class ProcessPsbtResponse(psbt: Psbt, complete: Boolean) {
+
+    import fr.acinq.bitcoin.psbt.UpdateFailure
+    import fr.acinq.bitcoin.scalacompat.KotlinUtils._
+
+    /** Transaction with all available witnesses. */
+    val partiallySignedTx: Transaction = {
+      var tx = psbt.global.tx
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala:136]: this seems to be using a version of bitcoin-kmp that has not been released yet, for now we have to keep using `getGlobal` instead

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala [hunk 2]
```diff
@@ -773,28 +773,30 @@ private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Respon
...
+      // We only sign our wallet inputs, and check that we can spend our wallet outputs.
+      val ourWalletInputs = unsignedTx.localInputs.map(i => tx.txIn.indexWhere(_.outPoint == i.outPoint)).filter(_ >= 0)
+      val ourWalletOutputs = unsignedTx.localOutputs.map {
         case Output.Local.Change(_, amount, pubkeyScript) => tx.txOut.indexWhere(output => output.amount == amount && output.publicKeyScript == pubkeyScript)
-      }
+        // Non-change outputs may go to an external address (typically during a splice-out).
+        // Here we only keep outputs which are ours i.e explicitly go back into our wallet.
+        // We trust that non-change outputs are valid: this only works if the entry point for creating such outputs is trusted (for example, a secure API call).
+        case _: Output.Local.NonChange => -1
+      }.filter(_ >= 0)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:784]: Same here: all local change outputs should be found in the tx ?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala [hunk 3]
```diff
@@ -343,9 +343,9 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
         }
         val sig = keyManager.sign(claimHtlcTx.txInfo, keyManager.htlcPoint(channelKeyPath), remotePerCommitmentPoint, TxOwner.Local, cmd.commitment.params.commitmentFormat)
         val signedTx = claimHtlcTx match {
-          case claimSuccess: ClaimHtlcSuccessWithWitnessData => claimSuccess.copy(txInfo = addSigs(claimSuccess.txInfo, sig, claimSuccess.preimage))
+          case ClaimHtlcSuccessWithWitnessData(txInfo, preimage) => ClaimHtlcSuccessWithWitnessData(addSigs(txInfo, sig, preimage), preimage)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:346]: Same as above

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala:136]: That's weird, how come it required it on my machine? I shouldn't have any override of the official `bitcoin-lib`, I'll look into it...

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala [hunk 4]
```diff
@@ -773,28 +773,30 @@ private class InteractiveTxBuilder(replyTo: ActorRef[InteractiveTxBuilder.Respon
...
-      // we only sign our wallet inputs, and check that we can spend our wallet outputs
-      val ourWalletInputs = unsignedTx.localInputs.map(i => tx.txIn.indexWhere(_.outPoint == OutPoint(i.previousTx, i.previousTxOutput.toInt)))
-      val ourWalletOutputs = unsignedTx.localOutputs.collect {
-        // README!! there are 2 types of local outputs:
-        // Change, which go back into our wallet
-        // NonChange, which go to an external address (typically during a splice-out)
-        // Here we only keep outputs which are ours i.e go back into our wallet
-        // And we trust that NonChange outputs are valid. This only works if the entry point for creating such outputs is trusted (for example, a secure API call)
+      // We only sign our wallet inputs, and check that we can spend our wallet outputs.
+      val ourWalletInputs = unsignedTx.localInputs.map(i => tx.txIn.indexWhere(_.outPoint == i.outPoint)).filter(_ >= 0)
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:777]: Good catch, I think you're right, there is no risk here that one of the inputs was removed. I'll look into it more.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala [hunk 5]
```diff
@@ -317,9 +317,9 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
   private def sign(fundedTx: ReplaceableTxWithWitnessData, txFeerate: FeeratePerKw, amountIn: Satoshi): Behavior[Command] = {
     val channelKeyPath = keyManager.keyPath(cmd.commitment.localParams, cmd.commitment.params.channelConfig)
     fundedTx match {
-      case claimAnchorTx: ClaimLocalAnchorWithWitnessData =>
-        val localSig = keyManager.sign(claimAnchorTx.txInfo, keyManager.fundingPublicKey(cmd.commitment.localParams.fundingKeyPath, cmd.commitment.fundingTxIndex), TxOwner.Local, cmd.commitment.params.commitmentFormat)
-        val signedTx = claimAnchorTx.copy(txInfo = addSigs(claimAnchorTx.txInfo, localSig))
+      case ClaimLocalAnchorWithWitnessData(anchorTx) =>
+        val localSig = keyManager.sign(anchorTx, keyManager.fundingPublicKey(cmd.commitment.localParams.fundingKeyPath, cmd.commitment.fundingTxIndex), TxOwner.Local, cmd.commitment.params.commitmentFormat)
+        val signedTx = ClaimLocalAnchorWithWitnessData(addSigs(anchorTx, localSig))
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:322]: My reasoning here is on the contrary, that it's safer to construct it entirely, because if you call `copy`, you cannot be sure that you didn't forget to also update another field in the object? Whereas recreating the object forces you to make sure all fields are correctly updated.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala [hunk 6]
```diff
@@ -125,4 +125,30 @@ object OnChainWallet {
   final case class FundTransactionResponse(tx: Transaction, fee: Satoshi, changePosition: Option[Int]) {
     val amountIn: Satoshi = fee + tx.txOut.map(_.amount).sum
   }
+
+  final case class ProcessPsbtResponse(psbt: Psbt, complete: Boolean) {
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala:129]: Because otherwise the dependency was wrong: `BitcoinCoreClient` extends `OnChainWallet`, but `OnChainWallet` defined functions that returned objects that were defined in `BitcoinCoreClient`...

The functions exposed in a trait should generally not return types that are only defined in one of the child classes implementing that trait, right?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala]: I removed the previous last commit that made those changes (https://github.com/ACINQ/eclair/pull/2726/commits/7eb74a9c6cfbee31cb2b0d47fbc946f6cbc8c53b), I don't remember how I got that invalid 0.13 version in my local maven repository!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Removed in https://github.com/ACINQ/eclair/pull/2726/commits/e25ce20eac87a0ccadd94d1975094382cc12f9ab

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Removed in https://github.com/ACINQ/eclair/pull/2726/commits/e25ce20eac87a0ccadd94d1975094382cc12f9ab

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/OnChainWallet.scala:129]: Right, and it's consistent with the definition of `FundTransactionResponse` for example.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:322]: I disagree: `.copy()` (which we use in our code base a lot) makes the code and its intent really clear: if a field should have been updated and was not it will be easy to spot, and that would be a bug. Creating a new objects could fail in ways that are not easy to spot (if there are multiple constructors with different parameters for example).

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:322]: > it will be easy to spot, and that would be a bug

I don't think it will be easy to spot at all?

Imagine for example that we add a field to `ClaimLocalAnchorWithWitnessData`. If we use `.copy`, we need to manually scan the whole codebase for every place where we copy instance of that field and verify that we update (or not) the new field. If we use a constructor, then the compiler automatically tells us and we have to decide for each object creation what values to put in that field.

If you're still not convinced by that argument I can change it, but I think it's a mistake...

[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:322]: Either way seems fine to me, as long as our policy is to not introduce fields with default values to existing classes.

\start(bikeshedding)
I generally like `.copy` because it makes it explicit what the intent is, rather then needing to parse the line visually for what changed; manually copying fields might also introduce an error as mentioned if a new field is added with a default value. However, I believe our policy is (or should be) to avoid adding fields with default values to existing classes.  In this case, the compiler will warn us everywhere the class is initiated and trigger any analysis of places it should change independent of when `.copy` is used. 
\end(bikeshedding)

**Review by remyers — APPROVED** [review]
Looks good to me; reviewing this PR was a good opportunity to improve my understanding of Scala code style. Otherwise I didn't see any issues introduced.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala:322]: It's true that the intent is clearer with `.copy`, anyway this part of the code is extensively tested so it should catch potential issues. I'll change it to `.copy`.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala]: Done in https://github.com/ACINQ/eclair/pull/2726/commits/9ff76f49b0a013329262e8b1ad6e10ef6aa88cfb

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala (hunk 3 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala]: Done in https://github.com/ACINQ/eclair/pull/2726/commits/9ff76f49b0a013329262e8b1ad6e10ef6aa88cfb

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala [hunk 7]
```diff
@@ -372,7 +374,7 @@ class BitcoinCoreClient(val rpcClient: BitcoinJsonRPCClient, val onchainKeyManag
       // we ask bitcoin core to add inputs to the funding tx, and use the specified change address
       FundTransactionResponse(tx, fee, _) <- fundTransaction(partialFundingTx, FundTransactionOptions(feerate))
       lockedUtxos = tx.txIn.map(_.outPoint)
-      signedTx <- unlockIfFails(tx.txid, lockedUtxos)(sign(tx, fee, feerate))
+      signedTx <- unlockIfFails(lockedUtxos)(verifyAndSign(tx, fee, feerate))
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:377]: When we used to log details about that, there were a lot of such failures in the logs, mostly when utxos disappear (because they're double-spent). If we fail to unlock a utxo, we can't do anything about it apart giving up, so why bother logging anything and making the code more complex?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala [hunk 8]
```diff
@@ -48,19 +48,24 @@ import scala.util.{Failure, Success, Try}
...
   import BitcoinCoreClient._
 
   implicit val formats: Formats = org.json4s.DefaultFormats
 
-  val useEclairSigner = onchainKeyManager_opt.exists(m => rpcClient.wallet.contains(m.wallet))
+  onChainKeyManager_opt.foreach { keyManager =>
+    require(rpcClient.wallet.contains(keyManager.wallet), s"eclair-backed bitcoin wallet mismatch: eclair-signer.conf uses wallet=${keyManager.wallet}, but eclair.conf uses wallet=${rpcClient.wallet.getOrElse("")}")
+  }
+
+  val useEclairSigner = onChainKeyManager_opt.nonEmpty
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:64]: Yes, that was explicitly mentioned in the commit message for https://github.com/ACINQ/eclair/pull/2726/commits/9263f40ad0a67e53a10a2801626f7ae517b04a8f

Supporting this case made the code very hard to follow, it wasn't tested and was quite bug-prone. I didn't know why that scenario in the first place, it really was hard to understand.

Can you provide more details about the exact flow you'd like to support, I'm still struggling to understand it? Then we'll see if we can support it in a simpler way than supporting this weird hybrid mode.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:64]: If I'm understanding you correctly, can't this simply be handled in `Setup.scala`? In that scenario the user wants his eclair to keep using a `BitcoinCoreClient` with `onChainKeyManager_opt = None`, don't they? And since they provided an `eclair-signer.conf`, they should be able to get the descriptors (maybe needs small code changes to `Setup.scala`)?

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:63]: I went further and moved the creation of the eclair-backed wallet to the on-chain key manager in https://github.com/ACINQ/eclair/pull/2726/commits/221d257f8ae5e2e231d1c6ca0fe9eab9b7788fcc

The `createEclairBackedWallet` flow had always bothered me, we were jumping through a few hops of indirection and implicit wallet name matching which made it hard to reason about. I believe the intent is clearer now, and it should support the scenario you're mentioning?

[the remark this answers is not in the document]
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:63]: this is still not functionally equivalent to what we had before: if users add an `eclair-signer.conf` to their node's data directory, then on startup eclair will try to create an empty wallet and import descriptors even if its `bitcoind.wallet` sill points to a regular bitcoin wallet. If the timestamp in `eclair-signer.conf` is too old,. this will fail, eclair will not start and users will not be able to use eclair's api to generate descriptors and import them manually.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:63]: Yes, that's true. But to be honest I think the issue is that it's really unclear to me what we are actually trying to support and why...if it was more clear what the spec actually is here, it would be easier to figure out what exactly to implement.

It looks like we're trying to support a lot of different use-cases that we're not going to use and users won't use either, which only makes the code more complex. Can we define the exact scope that we need instead and focus on only that?

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/BitcoinCoreClient.scala:63]: All the migration scenarios should now be supported with https://github.com/ACINQ/eclair/pull/2726/commits/eb49eba64bd69236bac8d1d1b8887a8b586e6598, let me know if you find any issue!

**Review by sstone — APPROVED** [review]
